File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder output = new StringBuilder();
	static StringTokenizer token;

	static int n, k;

	static int[] w, v;

	static int[][] arr;

	public static void main(String[] args) throws Exception {
		token = new StringTokenizer(input.readLine());

		n = Integer.parseInt(token.nextToken()); // 물건의 수
		k = Integer.parseInt(token.nextToken()); // 버틸수 있는 무게

		arr = new int[n + 1][k + 1];
		w = new int[n + 1];
		v = new int[n + 1];

		for (int i = 1; i <= n; i++) {
			token = new StringTokenizer(input.readLine());
			w[i] = Integer.parseInt(token.nextToken());
			v[i] = Integer.parseInt(token.nextToken());
		}

		for (int i = 1; i <= n; i++) { // 물건의 개수 만큼 돌리기
			for (int j = 1; j <= k; j++) { // 무게 0 부터 무게 k까지
				if (j - w[i] >= 0) {
					int a = arr[i - 1][j];
					int b = arr[i - 1][j - w[i]] + v[i];
					arr[i][j] = Math.max(a, b);
				} else {
					arr[i][j] = arr[i - 1][j];
				}
			}
		}

		System.out.println(arr[n][k]);

	}
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.kt
import java.lang.Integer.max

fun main() {
    val input = readln().split(" ").map { it.toInt() }
    val n = input[0]
    val k = input[1]

    val weight: MutableList<Int> = mutableListOf()
    val value: MutableList<Int> = mutableListOf()

    for (i in 0 until n) {
        val input = readln().split(" ").map { it.toInt() }
        weight.add(input[0])
        value.add(input[1])
    }

    val arr = Array(101) { IntArray(100001) }
    
    for (i in 1 until n + 1) {
        for (j in 1 until k + 1) {
            if (weight[i - 1] <= j) {
                arr[i][j] = max(arr[i - 1][j], arr[i - 1][j - weight[i - 1]] + value[i - 1])
            } else {
                arr[i][j] = arr[i - 1][j]
            }
        }
    }
    print(arr[n][k])
}
